File: front/src/api/friend-template.ts
import { Friend } from './friend'
import { withApiBase } from './base'

export interface FriendTemplate {
  id: number
  name: string
  avatar?: string | null
  description: string
  system_prompt: string
  initial_message?: string | null
  tags?: string[] | null
  created_at: string
  updated_at: string
}

export interface FriendTemplateQuery {
  page?: number
  size?: number
  tag?: string
  q?: string
}

export async function getFriendTemplates(query: FriendTemplateQuery = {}): Promise<FriendTemplate[]> {
  const params = new URLSearchParams()
  if (query.page) params.set('page', query.page.toString())
  if (query.size) params.set('size', query.size.toString())
  if (query.tag) params.set('tag', query.tag)
  if (query.q) params.set('q', query.q)

  const url = withApiBase(`/api/friend-templates/${params.toString() ? `?${params}` : ''}`)
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error('Failed to fetch friend templates')
  }
  return response.json()
}

export async function cloneFriendTemplate(id: number): Promise<Friend> {
  const url = withApiBase(`/api/friend-templates/${id}/clone`)
  const response = await fetch(url, {
    method: 'POST',
  })
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to clone friend' }))
    throw new Error(error.detail || 'Failed to clone friend')
  }
  return response.json()
}

export interface PersonaGenerateRequest {
  description: string
  name?: string
}

export interface PersonaGenerateResponse {
  name: string
  description: string
  system_prompt: string
  initial_message: string
}

export interface FriendTemplateCreateFriend {
  name: string
  avatar?: string | null
  description: string
  system_prompt: string
  initial_message?: string | null
}

/**
 * 根据描述自动生成 Persona 设定
 */
export async function getFriendTemplateTags(): Promise<string[]> {
  const url = withApiBase('/api/friend-templates/tags')
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error('Failed to fetch friend template tags')
  }
  return response.json()
}


export async function* generatePersonaStream(
  payload: PersonaGenerateRequest,
  options: { signal?: AbortSignal } = {}
): AsyncGenerator<{ event: string, data: any }> {
  const response = await fetch(withApiBase('/api/friend-templates/generate/stream'), {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
    signal: options.signal,
  })

  if (!response.ok) {
    throw new Error('Failed to generate persona stream')
  }

  const reader = response.body?.getReader()
  if (!reader) return

  const decoder = new TextDecoder()
  let buffer = ''

  while (true) {
    const { done, value } = await reader.read()
    if (done) {
      break
    }

    buffer += decoder.decode(value, { stream: true })
    const parts = buffer.split('\n\n')
    buffer = parts.pop() || ''

    for (const part of parts) {
      const lines = part.split('\n')
      let eventType = 'delta'
      let dataString = ''

      for (const line of lines) {
        if (line.startsWith('event: ')) {
          eventType = line.slice(7).trim()
        } else if (line.startsWith('data: ')) {
          dataString += (dataString ? '\n' : '') + line.slice(6)
        }
      }

      if (dataString) {
        try {
          const data = JSON.parse(dataString)
          yield { event: eventType, data }
        } catch (e) {
          console.error('Failed to parse SSE data JSON:', e)
          yield { event: eventType, data: dataString }
        }
      }
    }
  }
}

/**
 * 根据生成的设定直接创建好友
 */
export async function createFriendFromPayload(payload: FriendTemplateCreateFriend): Promise<Friend> {
  const url = withApiBase('/api/friend-templates/create-friend')
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to create friend' }))
    throw new Error(error.detail || 'Failed to create friend')
  }
  return response.json()
}

File: server/app/api/endpoints/friend_template.py
import json
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session

from app.api import deps
from app.schemas import friend as friend_schemas
from app.schemas import friend_template as friend_template_schemas
from app.schemas import persona_generator as persona_schemas
from app.services import friend_template_service
from app.services.persona_generator_service import persona_generator_service

router = APIRouter()


@router.get("/", response_model=List[friend_template_schemas.FriendTemplate])
def read_friend_templates(
    db: Session = Depends(deps.get_db),
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1),
    tag: Optional[str] = None,
    q: Optional[str] = None,
):
    return friend_template_service.get_friend_templates(
        db,
        page=page,
        size=size,
        tag=tag,
        q=q,
    )



@router.get("/tags", response_model=List[str])
def read_friend_template_tags(
    db: Session = Depends(deps.get_db),
):
    return friend_template_service.get_all_tags(db)


@router.post("/{template_id}/clone", response_model=friend_schemas.Friend)
def clone_friend_template(
    *,
    db: Session = Depends(deps.get_db),
    template_id: int,
):
    friend = friend_template_service.create_friend_from_template(db, template_id)
    if not friend:
        raise HTTPException(status_code=404, detail="Friend template not found")
    return friend


@router.post("/create-friend", response_model=friend_schemas.Friend)
def create_friend_from_template(
    *,
    db: Session = Depends(deps.get_db),
    payload: friend_template_schemas.FriendTemplateCreateFriend,
):
    return friend_template_service.create_friend_from_payload(db, payload)


@router.post("/generate/stream")
async def generate_persona_stream(
    *,
    db: Session = Depends(deps.get_db),
    payload: persona_schemas.PersonaGenerateRequest,
):
    """
    根据描述自动生成 Persona 设定（SSE 流式）
    """
    async def event_generator():
        async for event_data in persona_generator_service.generate_persona_stream(db, payload):
            event_type = event_data.get("event", "delta")
            data_payload = event_data.get("data", {})
            json_data = json.dumps(data_payload, ensure_ascii=False)
            yield f"event: {event_type}\ndata: {json_data}\n\n"

    return StreamingResponse(event_generator(), media_type="text/event-stream")
